accounts-db/src/tiered_storage/hot.rs

```diff
@@ -574,14 +575,30 @@ impl HotStorageWriter {
                 })
                 .unwrap_or((0, &OWNER_NO_OWNER, &[], false, None, None));
             let owner_offset = owners_table.insert(owner);
-            cursor += self.write_account(
+            let stored_size = self.write_account(
                 lamports,
                 owner_offset,
                 data,
                 executable,
                 rent_epoch,
                 account_hash,
             )?;
+            cursor += stored_size;
+
+            stored_infos.push(StoredAccountInfo {
+                // Here we pass the IndexOffset as the get_account() API
+                // takes IndexOffset.  Given the account address is also
+                // maintained outside the TieredStorage, a potential optimization
+                // is to store AccountOffset instead, which can further save
+                // one jump from the index block to the accounts block.
+                offset: index.len(),
+                // The size here might be slightly bigger than the actual
+                // storage size of one account as one owner address could be
+                // shared by multiple accounts.
+                size: stored_size
+                    + footer.index_block_format.entry_size::<HotAccountOffset>()
+                    + std::mem::size_of::<Pubkey>(),
```

Comment:
> We may want to tweak this API. It currently assumes that the total appended size is the same as summing up all the accounts appended. For Tiered Storage that's not the case.

Yep, I think there should exist a better API.  But if I remember it correctly, an estimation won't make accounts-db panic based on my previous run joining pop-net and main-net a while back.

> It'd probably be best to have this function return a tuple: a list of the stored account offsets, and a total size stored.

I think the size of each entry is used to estimate the saving of a shrink when selecting which file / account to shrink in accounts-db.  So I think it does want the information for each individual account, but no need to be super accurate.

> Since the size of the file in AccountStorageEntry::alive_bytes is used by shrink, I think we want to ensure we accurately compute the size here.

If I remember it correctly, accounts-db won't panic when hot-storage only provide an estimation based on my previous run on mainnet / popnet with the hot storage (a while back).

Reply:
Sure.  Let's do option 2.  Will create an issue as a follow-up.

Reply:
Just double-checked the accounts_db code.

When removing an account in AccountStorageEntry, it will maintain the `alive_bytes` by subtracting the stored_size.  In that case, I think it's safer to not include the owner here in TieredStorage otherwise the alive_bytes will go negative.

Will update the PR to use an estimate that will no larger than the actual stored size.

Reply:
Updated the PR to consider only stored size directly contribute by the account (i.e., mainly account meta, data, and address).

Reply:
As for the API, the stored size of each account is still used when computing the estimated storage size saving when shrinking a specific accounts file (or append-vec).  So I think the API should be good for now.

Reply:
Issue created: https://github.com/solana-labs/solana/issues/35041